packages/lesswrong/lib/analyticsEvents/ultraFeedEvents.ts

```diff
@@ -0,0 +1,130 @@
+import { useRef, useEffect, useCallback } from 'react';
+import { useMutation } from '@apollo/client';
+import gql from 'graphql-tag';
+
+interface UltraFeedEventInput {
+  eventType: 'viewed' | 'expanded';
+  documentId: string;
+  documentType: 'post' | 'comment' | 'spotlight';
+  postId?: string;
+  expansionLevel?: number;
+  maxExpansionReached?: boolean;
+  wordCount?: number;
+}
+
+const LOG_ULTRA_FEED_EVENT = gql`
+  mutation LogUltraFeedEvent($input: UltraFeedEventInput!) {
+    logUltraFeedEvent(input: $input) {
+      # Optionally return something, like the created event ID or just success
+      _id
+    }
+  }
+`;
+
+
+
+interface UseUltraFeedLoggingOptions {
+  documentId: string;
+  documentType: 'post' | 'comment' | 'spotlight';
+  postId?: string;
+  enabled?: boolean;
+}
+
+const VIEW_THRESHOLD_MS = 500; // Minimum time on screen to trigger 'viewed'
+
+export const useUltraFeedLogging = ({
+  documentId,
+  documentType,
+  postId,
+  enabled = true,
+}: UseUltraFeedLoggingOptions) => {
+  const [logEventMutation] = useMutation(LOG_ULTRA_FEED_EVENT);
+
+  const targetRef = useRef<HTMLDivElement | null>(null);
+  const viewedSentRef = useRef(false);
+  const timerRef = useRef<NodeJS.Timeout | null>(null);
+
+  const cleanUpObserver = useRef<() => void>(() => {});
+
+  useEffect(() => {
+    const currentElement = targetRef.current as Element | null;
+    if (!enabled || !currentElement || viewedSentRef.current) {
+      return;
+    }
+
+    const observer = new IntersectionObserver(
```

Comment:
like `UltraFeedObserver`, would benefit from an english-language description of what this is meant to be doing (and how they're related/unrelated)

Reply:
deleted file anyway